Sources/AsyncHTTPClient/AsyncAwait/Transaction.swift

```diff
@@ -0,0 +1,342 @@
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+import Logging
+import NIOConcurrencyHelpers
+import NIOCore
+import NIOHTTP1
+import NIOSSL
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+final class Transaction {
+    let logger: Logger
+
+    let request: HTTPClientRequest.Prepared
+
+    let connectionDeadline: NIODeadline
+    let preferredEventLoop: EventLoop
+    let requestOptions: RequestOptions
+
+    private let stateLock = Lock()
+    private var state: StateMachine
+
+    init(
+        request: HTTPClientRequest.Prepared,
+        requestOptions: RequestOptions,
+        logger: Logger,
+        connectionDeadline: NIODeadline,
+        preferredEventLoop: EventLoop,
+        responseContinuation: CheckedContinuation<HTTPClientResponse, Error>
+    ) {
+        self.request = request
+        self.requestOptions = requestOptions
+        self.logger = logger
+        self.connectionDeadline = connectionDeadline
+        self.preferredEventLoop = preferredEventLoop
+        self.state = StateMachine(responseContinuation)
+    }
+
+    func cancel() {
+        self.fail(HTTPClientError.cancelled)
+    }
+
+    // MARK: Request body helpers
+
+    private func writeOnceAndOneTimeOnly(byteBuffer: ByteBuffer) {
+        // This method is synchronously invoked after sending the request head. For this reason we
+        // can make a number of assumptions, how the state machine will react.
+        let writeAction = self.stateLock.withLock {
+            self.state.producedNextRequestPart(byteBuffer)
+        }
+
+        switch writeAction {
+        case .write(let part, let executor, continue: _):
+            // the continue flag can be ignored. We will maximum write the request's end.
+            executor.writeRequestBodyPart(.byteBuffer(part), request: self)
+            // we break here to continue in the function flow.
+
+        case .none:
+            // an error/cancellation has happened. we don't need to continue here
+            return
+        }
+
+        let finishAction = self.stateLock.withLock {
+            self.state.finishRequestBodyStream()
+        }
+
+        switch finishAction {
+        case .forwardStreamFinished(let executor):
+            executor.finishRequestBodyStream(self)
+        case .none:
+            // an error/cancellation has happened. we don't need to continue here
+            return
+        }
+    }
+
+    private func continueRequestBodyStream(
+        _ allocator: ByteBufferAllocator,
+        next: @escaping ((ByteBufferAllocator) async throws -> ByteBuffer?)
+    ) {
+        Task {
```

Comment:
Do we want to create new tasks here? Is there any reason not to use a single long async sequence and be able to translate `pause` into an `await` call?

Reply:
We could do this. However in this case we would have at least one allocation for every write pause, as we would need to create a new continuation. 

I guess in the NIO/Async translation layer the same rules should apply to Continuations that apply for EventLoopFutures: We should try to reduce them as much as possible.

Reply:
Just thinking about it, a new Task probably also causes a new allocation. So the issue comes down to: Do we want to allocate a continuation or a new Task for every pause.

Reply:
I guess the big pro for creating a new task for each continue here, is that we don't have to do a fastPath/slowPath approach here to not allocate a Continuation on every body part.

Reply:
Why would we have to allocate a new continuation on every body part? The same part of the code that tells us to wait can allocate the new continuation.